onnxruntime/core/providers/webnn/builders/impl/split_op_builder.cc

```diff
@@ -57,12 +55,18 @@ Status SplitOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
   axis = SafeInt<int32_t>(HandleNegativeAxis(axis, rank));
   options.set("axis", axis);
 
-  if (!GetTensorName(input_defs, 1).empty()) {
-    // Inputs contains optional 'split' input
-    std::vector<int32_t> splits;
-    const auto& initializers(model_builder.GetInitializerTensors());
-    const auto& split_tensor = *initializers.at(input_defs[1]->Name());
-    ORT_RETURN_IF_NOT(ReadIntArrayFrom1DTensor(split_tensor, splits, logger), "Cannot get split.");
+  // Inputs contains optional 'split' input or attribute.
+  if (!GetTensorName(input_defs, 1).empty() || node.SinceVersion() < 13) {
```

Comment:
🤔 This is becoming more complicated, and the flow is confusing because we're not checking things in any particular order, having multiple nested `if`s, are intermingling attribute reading with WebNN calls, and are writing conditions in terms of their *opposites* (e.g. `if (opset >= 13 && otherCondition)` is easier to process than `< 13`, and `tensor has a name` rather than if `not tensor has no name`). So if we arranged like so, it would be easier for future people to follow (and shorter too, but I might be missing some code 😉): e.g.

```c++
uint32_t split_count = 0;
std::vector<uint32_t> splits = helper.Get("split", std::vector<uint32_t>{});

// Read either the split count or explicit split lengths from the various attributes over opset versions.
if (helper.HasAttr("num_outputs")) {
    split_count = helper.Get("split", 0);
} else if (node.SinceVersion() >= 13 && TensorHasName(input_defs, 1)) {
    const auto& initializers(model_builder.GetInitializerTensors());
    const auto& split_tensor = *initializers.at(input_defs[1]->Name());
    ORT_RETURN_IF_NOT(ReadIntArrayFrom1DTensor(split_tensor, splits, logger), "Cannot get input for split.");
} else if (helper.HasAttr("split")) {
    splits = helper.Get("split", std::vector<uint32_t>{});
} else { // split into equal parts
    split_count = output_defs.size();
}

// Check that the splits evenly divide.
if (split_count > 0 && splits.empty() && input_shape[axis] % split_count != 0) {
  // Divide inputs into variable size outputs:
  splits.insert(splits.end(), split_count - 1, gsl::narrow<uint32_t>(input_shape[axis]) / split_count);
  splits.insert(splits.end(), gsl::narrow<uint32_t>(input_shape[axis]) % split_count);
}

if (splits.empty()) {
  output_array = model_builder.GetBuilder().call<emscripten::val>(
      "split", input, static_cast<int32_t>(split_count), options);
} else {
  output_array = model_builder.GetBuilder().call<emscripten::val>(
      "split", input, emscripten::val::array(splits), options);
}
```

Reply:
Thanks, that is clean and clear!

Reply:
🧼🧹 Thanks for the update.